src/accelerate/accelerator.py

```diff
@@ -857,7 +857,10 @@ def no_sync(self, model):
 
     def _do_sync(self):
         "Sets the right `sync_gradients` context and either resets or increases `self.step`"
-        if self.gradient_state.end_of_dataloader:
+        if (
```

Comment:
This still won't solve the issue. At the`end_of_dataloader`, `_set_sync_gradients` should always be True. The flag should only handle resetting of `self.step`

Reply:
With `transformers` we always set `sync_with_dataloader` to `False` so this will actually never happen, and it will only rely on the step

Reply:
```
# should this be under the accumulate context manager?
                # the `or` condition of `steps_in_epoch <= args.gradient_accumulation_steps` is not covered
                # in accelerate
                if total_batched_samples % args.gradient_accumulation_steps == 0 or (
                    # last step in epoch but step is always smaller than gradient_accumulation_steps
                    steps_in_epoch <= args.gradient_accumulation_steps
                    and (step + 1) == steps_in_epoch
                ):
```

When `steps_in_epoch <= args.gradient_accumulation_steps`, gradients should be synced at the end of dataloader as well as across epochs every gradient_accumulation_steps.

Reply:
E.g. epoch has 3 steps and `gradient_accumulation_steps` is 4. In this case, gradients should be synced at 3 step as that is the end of dataloader as well on 4th step (1st step of second epoch). This is to be on par with Trainer's behaviour.

Reply:
*just reread your first note*

Will adjust :)

Reply:
No, no, wait, even that would be incorrect for the steps_in_epoch > args.gradient_accumulation_steps. Let this PR be as is. I will explicitly set in Trainer the sync gradients to True for this peculiar case when `steps_in_epoch <= args.gradient_accumulation_steps`

Reply:
Yes, this is still required, I'll raise PR tomorrow using this

Reply:
Great! Okay. Let me know if we still need https://github.com/huggingface/transformers/pull/24405, otherwise I'll merge this for now thanks 🤗